File: mobile/src/api/beans.ts
import { authorizedFetch } from '@/src/lib/apiClient';
import { palette } from '@/src/theme';

export interface Bean {
  id: string;
  user_id: string;
  name: string;
  roaster?: string | null;
  origin?: string | null;
  process?: string | null;
  roast_level?: string | null;
  tasting_notes?: string | null;
  notes?: string | null;
  created_at: string;
  updated_at: string;
}

export interface BeanBody {
  name: string;
  roaster?: string;
  origin?: string;
  process?: string;
  roast_level?: string;
  tasting_notes?: string;
  notes?: string;
}

export const PROCESSES = [
  { id: 'washed',    label: 'Washed'    },
  { id: 'natural',   label: 'Natural'   },
  { id: 'honey',     label: 'Honey'     },
  { id: 'anaerobic', label: 'Anaerobic' },
  { id: 'other',     label: 'Other'     },
] as const;

export const ROAST_LEVELS = [
  { id: 'light',       label: 'Light',        color: palette.caramel400  },
  { id: 'medium_light',label: 'Medium light',  color: palette.caramel600  },
  { id: 'medium',      label: 'Medium',        color: palette.espresso500 },
  { id: 'medium_dark', label: 'Medium dark',   color: palette.espresso600 },
  { id: 'dark',        label: 'Dark',          color: palette.espresso800 },
] as const;

export function processLabel(id: string | null | undefined): string {
  return PROCESSES.find(p => p.id === id)?.label ?? '';
}

export function roastLabel(id: string | null | undefined): string {
  return ROAST_LEVELS.find(r => r.id === id)?.label ?? '';
}

export function roastColor(id: string | null | undefined): string {
  return ROAST_LEVELS.find(r => r.id === id)?.color ?? palette.cream500;
}

export const beansApi = {
  list:   ()                          => authorizedFetch<Bean[]>('/beans'),
  create: (body: BeanBody)            => authorizedFetch<Bean>('/beans',         { method: 'POST', body }),
  update: (id: string, body: BeanBody)=> authorizedFetch<Bean>(`/beans/${id}`,   { method: 'PUT',  body }),
  delete: (id: string)                => authorizedFetch<void>(`/beans/${id}`,   { method: 'DELETE' }),
};

File: mcp/beans.go
package main

import (
	"context"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type ListBeansParams struct{}

type CreateBeanParams struct {
	Name         string  `json:"name" jsonschema:"Bean name"`
	Roaster      *string `json:"roaster,omitempty" jsonschema:"Roaster name"`
	Origin       *string `json:"origin,omitempty" jsonschema:"Bean origin country or region"`
	Process      *string `json:"process,omitempty" jsonschema:"Processing method: washed, natural, honey, anaerobic, or other"`
	RoastLevel   *string `json:"roast_level,omitempty" jsonschema:"Roast level: light, medium_light, medium, medium_dark, or dark"`
	TastingNotes *string `json:"tasting_notes,omitempty" jsonschema:"Tasting notes"`
	Notes        *string `json:"notes,omitempty" jsonschema:"Additional notes"`
}

type UpdateBeanParams struct {
	ID           string  `json:"id" jsonschema:"Bean ID to update"`
	Name         string  `json:"name" jsonschema:"Bean name"`
	Roaster      *string `json:"roaster,omitempty" jsonschema:"Roaster name"`
	Origin       *string `json:"origin,omitempty" jsonschema:"Bean origin country or region"`
	Process      *string `json:"process,omitempty" jsonschema:"Processing method: washed, natural, honey, anaerobic, or other"`
	RoastLevel   *string `json:"roast_level,omitempty" jsonschema:"Roast level: light, medium_light, medium, medium_dark, or dark"`
	TastingNotes *string `json:"tasting_notes,omitempty" jsonschema:"Tasting notes"`
	Notes        *string `json:"notes,omitempty" jsonschema:"Additional notes"`
}

type DeleteBeanParams struct {
	ID string `json:"id" jsonschema:"Bean ID to delete"`
}

func registerBeanTools(server *mcp.Server, client *Client) {
	mcp.AddTool(server, &mcp.Tool{
		Name:        "list_beans",
		Description: "List all coffee beans for the authenticated user",
	}, func(ctx context.Context, req *mcp.CallToolRequest, params ListBeansParams) (*mcp.CallToolResult, any, error) {
		var result any
		if err := client.get(ctx, "/beans", &result); err != nil {
			return nil, nil, err
		}
		text, err := toJSON(result)
		if err != nil {
			return nil, nil, err
		}
		return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: text}}}, nil, nil
	})

	mcp.AddTool(server, &mcp.Tool{
		Name:        "create_bean",
		Description: "Create a new coffee bean entry",
	}, func(ctx context.Context, req *mcp.CallToolRequest, params CreateBeanParams) (*mcp.CallToolResult, any, error) {
		var result any
		if err := client.post(ctx, "/beans", params, &result); err != nil {
			return nil, nil, err
		}
		text, err := toJSON(result)
		if err != nil {
			return nil, nil, err
		}
		return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: text}}}, nil, nil
	})

	mcp.AddTool(server, &mcp.Tool{
		Name:        "update_bean",
		Description: "Update an existing coffee bean by ID",
	}, func(ctx context.Context, req *mcp.CallToolRequest, params UpdateBeanParams) (*mcp.CallToolResult, any, error) {
		body := struct {
			Name         string  `json:"name"`
			Roaster      *string `json:"roaster,omitempty"`
			Origin       *string `json:"origin,omitempty"`
			Process      *string `json:"process,omitempty"`
			RoastLevel   *string `json:"roast_level,omitempty"`
			TastingNotes *string `json:"tasting_notes,omitempty"`
			Notes        *string `json:"notes,omitempty"`
		}{
			Name:         params.Name,
			Roaster:      params.Roaster,
			Origin:       params.Origin,
			Process:      params.Process,
			RoastLevel:   params.RoastLevel,
			TastingNotes: params.TastingNotes,
			Notes:        params.Notes,
		}
		var result any
		if err := client.put(ctx, fmt.Sprintf("/beans/%s", params.ID), body, &result); err != nil {
			return nil, nil, err
		}
		text, err := toJSON(result)
		if err != nil {
			return nil, nil, err
		}
		return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: text}}}, nil, nil
	})

	mcp.AddTool(server, &mcp.Tool{
		Name:        "delete_bean",
		Description: "Delete a coffee bean by ID",
	}, func(ctx context.Context, req *mcp.CallToolRequest, params DeleteBeanParams) (*mcp.CallToolResult, any, error) {
		if err := client.delete(ctx, fmt.Sprintf("/beans/%s", params.ID)); err != nil {
			return nil, nil, err
		}
		return &mcp.CallToolResult{Content: []mcp.Content{&mcp.TextContent{Text: "Bean deleted successfully"}}}, nil, nil
	})
}
